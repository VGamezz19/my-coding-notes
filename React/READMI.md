# React

## React - Hooks

Nuevos en la version 16.8, Una forma de usar estado y otras funcionalidades de React sin escribir Clases.


## Porque han añadido los hooks?

Problemas con las clases:

- Componentes enormes

- Clases confusas

El problema principal esque React no nos proporciona el `Stateful primitive` si no trabajamos desde una class component.

La propuesta de los hooks no rompera las veriones anteriores, es un añadido.

## Hooks

### useState

UseState es uno de los hooks que nos proporciona la nueva version de React. No proporciona la posilbilidad de coger la funcionalidad `sateful` de react, en un funcional component.

 ```js
import React, { useState } from 'React';

export default function Todo ({text}) {
    const [name, setName]  = useState('Todo default text');

    //const name = // TodoDefault text
    //const setName =  // funciton set state name.

    function handleNameChange(e) {
        e.preventDefault();
        setName(e.target.value);
    }

    return (
        // HTML funcion component.
    );
}
```

Destructuramos las variables `name, setName`, donde *name* sera el valor actual del estado name `this.state.name`. la funcion *setName* sera la forma de modificar este espacion en memoria del estado, como si puera un `setState({})` pero más especifico.

Cuando llamamos el metodo que setea la propiedad, es cuando le decimos a React que necesitamos Renderizar el componente.