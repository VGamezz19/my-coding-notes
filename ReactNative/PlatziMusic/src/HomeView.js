/**
 * Sample React Native App
 * https://github.com/facebook/react-native
 * @flow
 */

import React, { Component } from 'react';
import {
  StyleSheet,
  View,
  ActivityIndicator
} from 'react-native';

import Icon from 'react-native-vector-icons/Ionicons'

import ArtistList from './ArtistList'
import { getArtists } from './api-client'

export default class HomeView extends Component {
  state = {
    artists: [],
    load: true
  }

  componentDidMount() {
    getArtists()
      .then(data => this.setState({ artists: data, load: false}))
  }

  render() {
    const artists = this.state.artists

    if(this.state.load) {
      contentHome = <ActivityIndicator size="large" color="#0000ff" />
    } else {
      contentHome = <ArtistList artists={artists} navigation ={this.props.navigation}/>
    }

    return (
      <View style={styles.container}>
        {contentHome}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'lightgray',
    paddingTop: 50,
  },
});