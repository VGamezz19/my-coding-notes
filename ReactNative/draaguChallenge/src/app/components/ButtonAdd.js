import React, { Component } from 'react';
import { Button } from 'react-native-elements'

export default class ButtonAdd extends Component{

    render() {
        return (
            <Button
                icon={{ name: 'code' }}
                backgroundColor='#03A9F4'
                buttonStyle={{ borderRadius: 0, marginLeft: 0, marginRight: 0, marginBottom: 0 }}
                title='LOAD MORE' />
        );
    }
}