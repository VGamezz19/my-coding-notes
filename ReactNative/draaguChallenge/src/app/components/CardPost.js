import React, { Component } from 'react';
import { Text } from 'react-native';
import { Card } from 'react-native-elements'

export default class CardPost extends Component {
  render() {
    return (
      <Card
        key={this.props.content.id}
        title={this.props.content.title}
        image={require('../../../public/img/defaultimg.jpg')}>
        <Text style={{ marginBottom: 10 }}>
          {this.props.content.body}
        </Text>
      </Card>
    );
  }
}