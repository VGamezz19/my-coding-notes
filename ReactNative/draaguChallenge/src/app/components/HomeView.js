import React, { Component } from 'react';
import {View,Text,Button,ListView,} from 'react-native';

import CardPost from './CardPost'

export default class HomeView extends Component {

  render() {
    return (
      <View style={{ flex: 1, alignItems: 'center', justifyContent: 'center' }}>
        <Text>Welcome Draagu Challeng</Text>
        <Button
          title="Go to ListView!"
          onPress={() => this.props.navigation.navigate('ListViewApp')}
        />
        <ListView
        enableEmptySections={true}
        dataSource={this.props.screenProps.dataSourceList}
        renderRow={post => {
          const postForCard = {
            id: post.id,
            title: post.title.rendered,
            body: post.excerpt.rendered
          }
          return <CardPost content={postForCard} />
        }
        } />
      </View>
    );
  }
}