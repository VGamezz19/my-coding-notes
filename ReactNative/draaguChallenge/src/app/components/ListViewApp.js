import React, { Component } from 'react';
import { StyleSheet, Text, View, ActivityIndicator, ListView, TouchableOpacity } from 'react-native'
import { apiDraagu } from '../../api/api-client'

import CardPost from './CardPost'
import ButtonAdd from './ButtonAdd'

let numPage = 1

export default class ListViewApp extends Component {
  constructor(props) {
    super(props);
  }

  getApiPost = () => {
    return apiDraagu().getPage(numPage).then(resPost => {
      this.props.screenProps.updateDataSource(resPost)
    })
  }

  onPressLoadMore = () => {
    this.props.screenProps.execLoad()
    numPage += 1
    this.getApiPost()
  }

  componentDidMount = () => this.getApiPost()

  render() {
    const load = this.props.screenProps.load
    let contentList

    if (load) {
      contentList = <ActivityIndicator size="large" color="#0000ff" />
    } else {
      contentList = <ListView
        enableEmptySections={true}
        dataSource={this.props.screenProps.dataSourceList}
        renderRow={post => {
          const postForCard = {
            id: post.id,
            title: post.title.rendered,
            body: post.excerpt.rendered
          }
          return <CardPost content={postForCard} />
        }
        } />
    }

    return (
      <View style={styles.container}>

        {contentList}

        {load ? undefined : (
          <TouchableOpacity
            style={styles.button}
            onPress={this.onPressLoadMore}>
            <Text> Load More </Text>
          </TouchableOpacity>
        )}
      </View>
    );
  }
}

//================================
//Styles
const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    paddingHorizontal: 10
  },
  button: {
    alignItems: 'center',
    backgroundColor: '#DDDDDD',
    padding: 10
  }
})