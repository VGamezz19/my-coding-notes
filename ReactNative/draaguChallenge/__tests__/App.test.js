import 'react-native';
import React from 'react';
import { ListView } from 'react-native'
import ButtonAdd from '../src/app/components/ButtonAdd'
import CardPost from '../src/app/components/CardPost'
import HomeView from '../src/app/components/HomeView'
import ListViewApp from '../src/app/components/ListViewApp'
//Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import "isomorphic-fetch"

it('render correctly component ButtonAdd', () => {
  const button = renderer.create(
    <ButtonAdd />
  );
});
it('rendering component CardPost', () => {
  const postForCard = {
    id: 1,
    title: "testeando Card",
    body: "texto card body"
  }

  const card = renderer.create(
    <CardPost content={postForCard} />
  );

});

it('rendering component HomeView', () => {
  const home = renderer.create(
    <HomeView screenProps={{
      dataSourceList: new ListView.DataSource({
        rowHasChanged: (r1, r2) => {
          r1 !== r2
        }
      })
    }} />
  );

});

it('rendering component ListViewApp', () => {
  const list = renderer.create(
    <ListViewApp screenProps={{
      dataSourceList: new ListView.DataSource({
        rowHasChanged: (r1, r2) => {
          r1 !== r2
        }
      })
    }} />
  );
  expect(list).toMatchSnapshot();
});