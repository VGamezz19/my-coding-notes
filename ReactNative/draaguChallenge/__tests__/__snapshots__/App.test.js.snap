// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering component ListViewApp 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 10,
    }
  }
>
  <RCTScrollView
    dataSource={
      ListViewDataSource {
        "items": "unknown",
      }
    }
    enableEmptySections={true}
    renderRow={[Function]}
    renderScrollComponent={[Function]}
  >
    <View />
  </RCTScrollView>
  <View
    accessibilityComponentType={undefined}
    accessibilityLabel={undefined}
    accessibilityTraits={undefined}
    accessible={true}
    collapsable={undefined}
    hasTVPreferredFocus={undefined}
    hitSlop={undefined}
    isTVSelectable={true}
    nativeID={undefined}
    onLayout={undefined}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#DDDDDD",
        "opacity": 1,
        "padding": 10,
      }
    }
    testID={undefined}
    tvParallaxProperties={undefined}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
       Load More 
    </Text>
  </View>
</View>
`;
